### PR TITLE
Register SGX enclave workloads automatically after loading enclave

### PR DESCRIPTION
**Pull request by manojgop**
- Added workload processor base class and prototype design pattern
  for automatic registration of workloads after loading enclave.
- Echo and hearbeat workloads will auto register.
- Enabled logging functionality in enclave to view SGX workload
  registration messages in stdout.
- Removed obsolete files related to workload handling.

Signed-off-by: Manoj Gopalakrishnan <manoj.gopalakrishnan@intel.com>

### CODE REVIEW

File: tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt [hunk 1]
```diff
@@ -171,6 +171,11 @@ ELSE()
   ADD_DEFINITIONS(-DTCF_DEBUG_BUILD=0)
 ENDIF()
 
+# Enable enclave logs
+IF(DEFINED ENV{TCF_ENCLAVE_LOG} AND "$ENV{TCF_ENCLAVE_LOG}" STREQUAL "1")
+  ADD_DEFINITIONS(-DTCF_ENCLAVE_LOG=1)
```
Comment (EugeneYYY) [inline tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt]: From security perspective no logs are allowed in the release build so we need to modify our definitions to have 
TCF_SGX_SYM - to build TCF in SGX simulation mode
TCF_DEBUG_BUILD - to build  for either HW or simulation mode
TCF_ENCLAVE_LOG - allows logging, but only if TCF_DEBUG_BUILD is defined, e.g.
IF(DEFINED ENV{TCF_ENCLAVE_LOG} AND "$ENV{TCF_ENCLAVE_LOG}" STREQUAL "1" AND TCF_DEBUG_BUILD} AND "$ENV{TCF_DEBUG_BUILD}" STREQUAL "1")

We must have a single clearly defined macro that turns off all debug support presenting a potential security issues

File: tc/sgx/trusted_worker_manager/enclave/enclave_utils.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ void Log(int level, const char* fmt, ...) {
     va_start(ap, fmt);
     vsnprintf_s(buf, BUFSIZ, fmt, ap);
     va_end(ap);
-#ifdef DEBUG
+#ifdef TCF_ENCLAVE_LOG
```
Comment (EugeneYYY) [inline tc/sgx/trusted_worker_manager/enclave/enclave_utils.cpp]: See comment above about preventing logs in the release build. It makes sense to change this macro to be 
#ifdef DEBUG
    #ifdef TCF_ENCLAVE_LOG
         ocall_Log(level, buf);
    #endif // TCF_ENCLAVE_LOG
#endif  // DEBUG

File: tc/sgx/trusted_worker_manager/enclave/work_order_processor.cpp [hunk 3]
```diff
@@ -229,11 +228,12 @@ namespace tcf {
...
             }
 
-            WorkloadProcessor processor;
             // Convert workload_id from hex string to string
             ByteArray workload_bytes = HexStringToBinary(workload_id);
             std::string workload_type(workload_bytes.begin(), workload_bytes.end());
-            processor.ProcessWorkOrder(
+            WorkloadProcessor *processor = 
+                WorkloadProcessor::CreateWorkloadProcessor(workload_type);
+            processor->ProcessWorkOrder(
```
Comment (EugeneYYY) [inline tc/sgx/trusted_worker_manager/enclave/work_order_processor.cpp]: To be on safe side, especially, if CreateWorkloadProcessor() implemenentation changes in the future, add a check
if ( processor == nullptr)
{
     ... through an exception
}

File: common/sgx_workload/workload_processor.cpp [hunk 4]
```diff
@@ -13,42 +13,46 @@
...
+   if (itr == workload_processor_table.end()) {
+       tcf::error::RuntimeError("Workload Processor not found in table");
+       return nullptr;
+   } else {
+       processor = (*itr).second;
+   }
+
+   // Clone the workload processor and return it.
+   if( processor == nullptr) {
+       return nullptr;
```
[the remark this answers is not in the document]
Comment (manojgop) [reply common/sgx_workload/workload_processor.cpp]: ok

File: tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt (hunk 1 above)
Comment (manojgop) [reply tc/sgx/trusted_worker_manager/enclave/CMakeLists.txt]: Currently TCF_DEBUG_BUILD is used to enable debugging symbols (-g option in CMakeList.txt). But I was thinking logging (via ocall) is independent of enabling debugging symbols. So you are suggesting to enable logs we have to have debug symbols ?

File: tc/sgx/trusted_worker_manager/enclave/work_order_processor.cpp (hunk 3 above)
Comment (manojgop) [reply tc/sgx/trusted_worker_manager/enclave/work_order_processor.cpp]: ok

**Comment by danintel** [comment]
These are my notes for this PR.

This PR allows a workload to be automatically registered.

Two macros defined in workload_processor.h:
**REGISTER_WORKLOAD_PROCESSOR**
```#define REGISTER_WORKLOAD_PROCESSOR(WORKLOADID_STR,TYPE) \
   WorkloadProcessor* TYPE##_myProcessor = \
      WorkloadProcessor::RegisterWorkloadProcessor(WORKLOADID_STR, new TYPE());```

Used to register a workload processor for a specific application.
Associates a string with a worker class.

Examples in examples/.../echo.cpp and heart_disease_evaluation.cpp
// echo.cpp:
    `REGISTER_WORKLOAD_PROCESSOR("echo-result",EchoResult)`
// heart_disease_evaluation.cpp:
    `REGISTER_WORKLOAD_PROCESSOR("heart-disease-eval",HeartDiseaseEval)`


**IMPL_WORKLOAD_PROCESSOR_CLONE**

```#define IMPL_WORKLOAD_PROCESSOR_CLONE(TYPE) \
   WorkloadProcessor* Clone() const { return new TYPE(*this); }```

Clones a an instance of class WorkloadProcessor for a worker.

Examples of usage:
// echo.h:
    `IMPL_WORKLOAD_PROCESSOR_CLONE(EchoResult)`
// heart_disease_evaluation.h:
    `IMPL_WORKLOAD_PROCESSOR_CLONE(HeartDiseaseEval)`

**Pull request merged**